inc/zoo/swar/SWAR.h

```diff
@@ -83,99 +88,153 @@ namespace detail {
 
     template<bool> struct Selector_impl {
         template<int Level>
-        constexpr static uint64_t execute(uint64_t v) {
+        constexpr static uint64_t execute(uint64_t v) noexcept {
             return popcount_logic<Level>(v);
         }
     };
     template<> struct Selector_impl<true> {
         template<int Level>
-        constexpr static uint64_t execute(uint64_t v) {
+        constexpr static uint64_t execute(uint64_t v) noexcept {
             return popcount_builtin<Level - 2>(v);
         }
     };
 
 }
 
 template<int Level>
-constexpr uint64_t popcount(uint64_t a) {
+constexpr uint64_t popcount(uint64_t a) noexcept {
     return detail::Selector_impl<2 < Level>::template execute<Level>(a);
 }
 
-static_assert(0x210 == popcount<0>(0x320), "");
-static_assert(0x4321 == popcount<1>(0xF754), "");
-static_assert(0x50004 == popcount<3>(0x3E001122), "");
-
 
-template<typename T> constexpr typename std::make_unsigned<T>::type msb_index(T v) {
+template<typename T> constexpr typename std::make_unsigned<T>::type msb_index(T v) noexcept {
     return 8*sizeof(T) - 1 - __builtin_clzll(v);
 }
 
-template<int Size, typename T = uint64_t> struct SWAR {
+template<typename T> constexpr typename std::make_unsigned<T>::type lsb_index(T v) noexcept {
+    return __builtin_ctzll(v) + 1;
+}
+
+template<int NBits, typename T = uint64_t> struct SWAR {
     SWAR() = default;
     constexpr explicit SWAR(T v): m_v(v) {}
+    constexpr explicit operator T() const noexcept { return m_v; }
 
     constexpr T value() const noexcept { return m_v; }
 
-    constexpr SWAR operator|(SWAR o) { return SWAR(m_v | o.m_v); }
-    constexpr SWAR operator&(SWAR o) { return SWAR(m_v & o.m_v); }
-    constexpr SWAR operator^(SWAR o) { return SWAR(m_v ^ o.m_v); }
+    constexpr SWAR operator|(SWAR o) const noexcept { return SWAR(m_v | o.m_v); }
+    constexpr SWAR operator&(SWAR o) const noexcept { return SWAR(m_v & o.m_v); }
+    constexpr SWAR operator^(SWAR o) const noexcept { return SWAR(m_v ^ o.m_v); }
 
-    constexpr bool operator==(SWAR o) { return m_v == o.m_v; }
-
-    constexpr T at(int position) {
-        constexpr auto filter = (T(1) << Size) - 1;
-        return filter & (m_v >> (Size * position));
+    constexpr T at(int position) const noexcept {
+        constexpr auto filter = (T(1) << NBits) - 1;
+        return filter & (m_v >> (NBits * position));
```

Comment:
I believe we will run into issues if we keep using `NBits * position` because when it is a multiplication, it is painfully expensive.  We ought to create the cheapo version as a partial template specialization for the case of powers of 2

Reply:
I'm not sure we're actually using at anywhere yet, but I do agree it is too expensive.  I was planning to ignore the problem until I thought of something smarter.

Reply:
What I mean is we will provide the API for the powers of 2 so that the user unambiguously knows it is using the fast and to prevent the mistake when not.  However, we ought to keep `at` in the way we have it now for generality